store/block_store_pruner.go

```diff
@@ -0,0 +1,168 @@
+// The source DB path should specify the parent directory within which blockstore.db can be found.
+// The minimum height should specify the oldest block that must be retained in the new blockstore.db,
+// it should be set to match the app height to ensure that the app can start processing blocks from
+// that height.
+func PruneBlockStore(srcDBPath string, cfg *BlockStoreConfig, minHeight int64) error {
+	srcDB, err := dbm.NewGoLevelDBWithOpts("blockstore", srcDBPath, &opt.Options{ReadOnly: true})
+	if err != nil {
+		return err
+	}
+	defer srcDB.Close()
+
+	srcBlockStore := blockchain.NewBlockStore(srcDB)
+
+	destDB := dbm.NewDB("pruned_blockstore", "leveldb", srcDBPath)
+	defer destDB.Close()
+
+	latestHeight := blockchain.LoadBlockStoreStateJSON(srcDB).Height
+	var targetHeight int64
+	targetHeight = latestHeight - cfg.NumBlocksToRetain
+	if minHeight < targetHeight {
+		targetHeight = minHeight
+	}
+	graceBlocks := (cfg.PruneGraceFactor / 100) * cfg.NumBlocksToRetain
+	oldestHeight := int64(-1)
+	// Find the oldest block
+	it := srcDB.Iterator(calcBlockMetaPrefix, prefixRangeEnd(calcBlockMetaPrefix))
+	for ; it.Valid(); it.Next() {
+		oldestBlockMetaKey := it.Key()
+		oldestHeight = getHeightFromKey(oldestBlockMetaKey)
+		break
+	}
+	// Number of blocks to prune is less than grace blocks then skip pruning
+	if ((targetHeight - oldestHeight) + 1) < graceBlocks {
+		return nil
+	}
+	// If minimum height is greater than target height, there are no blocks to prune, skip pruning
+	if oldestHeight >= targetHeight {
+		return nil
+	}
+
+	log.Info("[Block Store Pruner] Copying blocks", "fromHeight", targetHeight, "toHeight", latestHeight)
+
+	batch := destDB.NewBatch()
+	numBlocksWritten := int64(0)
+	for height := targetHeight; height <= latestHeight; height++ {
+		// skip if block metadata is not found
+		if !srcDB.Has(calcBlockMetaKey(height)) {
+			log.Info("[Block Store Pruner] Missing block", "height", height)
+			continue
+		}
+
+		meta := srcBlockStore.LoadBlockMeta(height)
+		for i := 0; i < meta.BlockID.PartsHeader.Total; i++ {
+			blockPartKey := calcBlockPartKey(height, i)
+			batch.Set(blockPartKey, srcDB.Get(blockPartKey))
+		}
+
+		blockMetaKey := calcBlockMetaKey(height)
+		blockCommitKey := calcBlockCommitKey(height - 1)
+		SeenCommitKey := calcSeenCommitKey(height)
+
+		batch.Set(blockMetaKey, srcDB.Get(blockMetaKey))
+		batch.Set(blockCommitKey, srcDB.Get(blockCommitKey))
+		batch.Set(SeenCommitKey, srcDB.Get(SeenCommitKey))
+
+		if numBlocksWritten%cfg.BatchSize == 0 {
+			batch.Write()
+			batch = destDB.NewBatch()
+		}
+		numBlocksWritten++
+	}
+
+	batch.Set(blockStoreKey, srcDB.Get(blockStoreKey))
+
+	if numBlocksWritten > 0 {
+		batch.WriteSync()
+	}
+
+	srcDB.Close()
+	destDB.Close()
+
+	log.Info("[Block Store Pruner] Finished copying blocks", "count", numBlocksWritten)
+
+	if numBlocksWritten > 0 {
+		// Rename original blockstore to blockstore.db.bak{N}
+		backupPath := getBackupDBPath(srcDBPath)
```

Comment:
I don't think we should keep the old blockstores. It's going to be useless and requires us to manually remove it later. We should just remove it here I think.